### ISSUE TITLE
@Qualifier annotations applied to a @Component or @Subcomponent type don't report an error

### ISSUE THREAD

**Issue by cgruber** [google/dagger#854]
When given something like:

```
@Qualifier
@interface FooScope {} // misnamed and confusing, but not really a scope
```

attached to a component or subcomponent like so:

```
@FooScope
@Component
interface FooComponent { ... }
```

the processor fails to note the incoherent usage, and compiles successfully, despite there being no behavior backing this mis-configuration. 

Failing test (for both subcomponent and component) in cl/166399791 internally, more or less replicating the above sample but with real code.

**Comment by ronshapiro** [comment]
Should this be an error-prone check? This is also funky:

```
@FooScopeButActuallyQualifier
class Foo {
  @Inject Foo() {}
}
```

**Comment by cgruber** [comment]
Error prone makes sense, but we can't guarantee that non-google users will be using error-prone, and dagger passes over all of these, so I'd make those dagger errors.

**Comment by cgruber** [comment]
But I agree that it's also an incoherent case, given the current spec. 

Thinking out loud a bit, I could see it being a feature on an `@Inject` constructor-annotated type - that is, the default implicit binding is qualified. That would be good for documenting cases like

```
@DefaultFoo
class Foo { 
  @Inject Foo() { ... }
  protected void someMethod() { /* default null behavior */ }
}
```

... but it could also be argued that this is not the greatest design, and that having a Foo interface, and a DefaultFoo type is better.  

The spec is agnostic to the above, saying: *While this specification covers applying qualifiers to fields and parameters only, some injector configurations might use qualifier annotations in other places (on methods or classes for example).*

We use it for component methods, so we're spec-variant, though not incompatible.

In any case, the two component-type cases are decidedly the wrong usage.  I would recommend we also prohibit the injectable-type class-annotation usage until and unless we make that a feature.

**Comment by ronshapiro** [comment]
SGTM. We should also provide a warning to "scope-sounding-names" that are `@Qualifier`'d

**Comment by cgruber** [comment]
Yeah.  Agreed. Not sure what the proper heuristic should be there, but I'm
sure there are some.low hanging fruit.
